### PR TITLE
add ndv5 algo files and azure vm detection logic

### PR DESCRIPTION
**Pull request by Andyli1007**

### CODE REVIEW

File: src/misc/msccl/msccl_lifecycle.cc [hunk 1]
```diff
@@ -77,9 +80,52 @@ static const char* mscclSchedulerPathEnv = "MSCCL_SCHEDULER";
 static const char* mscclSchedulerDefaultPath = "libmsccl-scheduler.so";
 static const char* mscclAlgoDirEnv = "MSCCL_ALGO_DIR";
 static const char* mscclAlgoDefaultDir = "msccl-algorithms";
+static const char* azureVMDetectionAgent = "http://169.254.169.254/metadata/instance?api-version=2019-06-04";
```
Comment (yzygitzh) [inline src/misc/msccl/msccl_lifecycle.cc:83]: Change to mscclAzureVMDetectionAgent to keep aligned?

File: src/misc/msccl/msccl_lifecycle.cc [hunk 2]
```diff
@@ -77,9 +80,52 @@ static const char* mscclSchedulerPathEnv = "MSCCL_SCHEDULER";
...
 extern "C" bool mscclUnitTestMode() __attribute__((__weak__));
 static const char* mscclUnitTestAlgoDefaultDir = "msccl-unit-test-algorithms";
 
+static size_t WriteCallback(void* contents, size_t size, size_t nmemb, std::string* userp) {
```
Comment (yzygitzh) [inline src/misc/msccl/msccl_lifecycle.cc:87]: WriteCallback -> writeCallback?

File: src/misc/msccl/msccl_lifecycle.cc [hunk 3]
```diff
@@ -77,9 +80,52 @@ static const char* mscclSchedulerPathEnv = "MSCCL_SCHEDULER";
...
+            vmSize = json["compute"]["vmSize"].get<std::string>();
+        }
+        curl_easy_cleanup(curl);
+    }
+    curl_global_cleanup();
+    if (vmSize.find("ND") != std::string::npos && vmSize.find("A100") != std::string::npos) {
+      updatedAlgoDir.append("/ndv4");
+    }
+    else if (vmSize.find("ND") != std::string::npos && vmSize.find("H100") != std::string::npos) {
+      updatedAlgoDir.append("/ndv4");
```
Comment (yzygitzh) [inline src/misc/msccl/msccl_lifecycle.cc:124]: Should be ndv5?

File: src/misc/msccl/msccl_lifecycle.cc (hunk 1 above)
Comment (Andyli1007) [reply src/misc/msccl/msccl_lifecycle.cc]: agree. Already regulated.

File: src/misc/msccl/msccl_lifecycle.cc (hunk 2 above)
Comment (Andyli1007) [reply src/misc/msccl/msccl_lifecycle.cc]: fixed

File: src/misc/msccl/msccl_lifecycle.cc (hunk 3 above)
Comment (Andyli1007) [reply src/misc/msccl/msccl_lifecycle.cc]: fixed.

**Review by yzygitzh — APPROVED** [review]

**Pull request merged**